File: src/main/java/net/miarma/mkernel/commands/tp/TpDenyCommand.java
package net.miarma.mkernel.commands.tp;

import java.util.List;
import java.util.Optional;

import org.bukkit.Bukkit;
import org.bukkit.entity.Player;

import dev.jorel.commandapi.CommandAPICommand;
import net.miarma.mkernel.common.minecraft.teleport.TpaRequest;
import net.miarma.mkernel.common.minecraft.teleport.TpaRequests;
import net.miarma.mkernel.config.CommandWrapper;
import net.miarma.mkernel.config.providers.CommandProvider;
import net.miarma.mkernel.config.providers.MessageProvider;
import net.miarma.mkernel.util.MessageUtil;

public class TpDenyCommand {
    public static void register() {
        CommandWrapper tpDenyCmd = CommandProvider.getTpDenyCommand();
        new CommandAPICommand(tpDenyCmd.getName())
            .withOptionalArguments(CommandProvider.Arguments.playersOptArg())
            .withPermission(tpDenyCmd.getPermission().base())
            .withFullDescription(tpDenyCmd.getDescription())
            .withShortDescription(tpDenyCmd.getDescription())
            .withUsage(tpDenyCmd.getUsage())
            .executesPlayer((sender, args) -> {
                if(args.count() == 0) {
                    Optional<TpaRequest> optTpaRequest = TpaRequests.getInstance().getRequests().stream()
                            .filter(r -> r.from().equals(sender))
                            .findFirst();

                    Optional<TpaRequest> optTpaHereRequest = TpaRequests.getInstance().getRequests().stream()
                            .filter(r -> r.to().equals(sender))
                            .findFirst();

                    if(optTpaRequest.isEmpty() && optTpaHereRequest.isEmpty()) {
                        MessageUtil.sendMessage(sender, MessageProvider.Errors.noRequestFound(), true);
                        return;
                    }

                    TpaRequest request;
                    request = optTpaRequest.orElseGet(optTpaHereRequest::get);
                    TpaRequests.getInstance().removeRequest(request);

                    Player target = Bukkit.getPlayer(request.from().getName());

                    MessageUtil.sendMessage(sender, tpDenyCmd.getMessages()[0], true,
                            List.of("%target%"), List.of(target.getName()));
                    MessageUtil.sendMessage(target, tpDenyCmd.getMessages()[1], true,
                            List.of("%sender%"), List.of(sender.getName()));

                } else {
                    Player target = Bukkit.getPlayer(args.getRaw(0));

                    if (target == null || !target.isOnline()) {
                        MessageUtil.sendMessage(sender, MessageProvider.Errors.playerNotFound(), true);
                        return;
                    }

                    TpaRequest request;
                    if(TpaRequests.getInstance().getTpaRequest(target, sender) != null) {
                        request = TpaRequests.getInstance().getTpaRequest(target, sender);
                    } else {
                        request = TpaRequests.getInstance().getTpaHereRequest(sender, target);
                    }

                    if (request == null) {
                        MessageUtil.sendMessage(sender, MessageProvider.Errors.noRequestFound(), true);
                        return;
                    }

                    TpaRequests.getInstance().removeRequest(request);

                    MessageUtil.sendMessage(sender, tpDenyCmd.getMessages()[0], true,
                            List.of("%target%"), List.of(target.getName()));
                    MessageUtil.sendMessage(target, tpDenyCmd.getMessages()[1], true,
                            List.of("%sender%"), List.of(sender.getName()));
                }
            })
            .register();
    }
}

File: src/main/kotlin/net/miarma/mkernel/command/tp/TpDenyCommand.kt
package net.miarma.mkernel.command.tp

import com.google.inject.Inject
import com.google.inject.Singleton
import dev.jorel.commandapi.arguments.PlayerProfileArgument
import dev.jorel.commandapi.kotlindsl.commandAPICommand
import dev.jorel.commandapi.kotlindsl.playerExecutor
import net.miarma.mkernel.common.service.impl.ConfigService
import net.miarma.mkernel.common.service.impl.MessageService
import net.miarma.mkernel.common.service.impl.TeleportService
import net.miarma.mkernel.common.teleport.TpaRequest
import net.miarma.mkernel.util.PlayerUtil

@Singleton
class TpDenyCommand @Inject constructor(
    private val configService: ConfigService,
    private val messageService: MessageService,
    private val teleportService: TeleportService
) {
    fun register() {
        commandAPICommand(configService.getString("commands.tpdeny.name")) {
            withOptionalArguments(PlayerProfileArgument(configService.getString("arguments.player")))
            withPermission(configService.getString("commands.tpdeny.permission"))
            withFullDescription(configService.getString("commands.tpdeny.description"))
            withUsage(configService.getString("commands.tpdeny.usage"))
            playerExecutor { sender, args ->
                if (args[0] == null) {
                    teleportService.getIncomingTpaRequest(sender) { request ->
                        if (request == null) {
                            messageService.builder(configService.getString("language.errors.noRequestFound")).withPrefix().send(sender)
                            return@getIncomingTpaRequest
                        }
                        denyRequest(sender, request.from, request)
                    }
                } else {
                    val target = PlayerUtil.fromArg(args[0])
                    if (target == null || !target.isOnline) {
                        messageService.builder(configService.getString("language.errors.playerNotFound")).withPrefix().send(sender)
                        return@playerExecutor
                    }

                    teleportService.getTpaRequest(target, sender) { request ->
                        if (request == null) {
                            messageService.builder(configService.getString("language.errors.noRequestFound")).withPrefix().send(sender)
                            return@getTpaRequest
                        }
                        denyRequest(sender, target, request)
                    }
                }
            }
        }
    }

    private fun denyRequest(sender: org.bukkit.entity.Player, target: org.bukkit.entity.Player, request: TpaRequest) {
        teleportService.removeRequest(request)
        messageService.builder(configService.getString("commands.tpdeny.messages.deniedToTarget")).withPrefix().forPlayer(target).tag("sender", sender.name).send(target)
        messageService.builder(configService.getString("commands.tpdeny.messages.denied")).withPrefix().tag("target", target.name).send(sender)
    }
}